### ISSUE TITLE
Anisotropy distribution

### ISSUE THREAD

**Issue by MikhailRyazanov** [PyAbel/PyAbel#254]
I did not fully read the previous “anisotropy” issues (#60 and #176), but from a more recent [discussion](https://github.com/PyAbel/PyAbel/issues/225#issuecomment-422654901) with @stggh, it seems that the means to calculate the anisotropy distribution (radially dependent anisotropy parameters) are still demanded and not really available.

So I have a [proof of concept](https://gist.github.com/MikhailRyazanov/8092647cbf3a4d5efc24e5120d364fd6) implementing several approaches that seem reasonable for fast calculations. They all seem to work, with somewhat different properties:

For the simple image with separate peaks (1 quadrant):
![img](https://user-images.githubusercontent.com/12854133/58145227-956f5a80-7c0e-11e9-8ad4-5f3a0f2d31ba.png)

![aniso](https://user-images.githubusercontent.com/12854133/58145032-ee8abe80-7c0d-11e9-8765-2d52b12751cc.png)

For the Dribinski sample image:
![drib](https://user-images.githubusercontent.com/12854133/58145135-45909380-7c0e-11e9-96e1-c383f9399369.png)

The implemented methods are:
1. `nearest` bins image pixels to the nearest radial bin.
2. `linear` bins image pixels, distributing them over the 2 closest radial bins with linear weights.
3. `remap` converts the image to a uniform polar grid (like `reproject_image_into_polar`).

Except these differences, they all do linear least-squares fitting to cos<sup>_l_</sup> θ terms, then convert them to Legendre polynomial and finally to the intensity and anisotropy parameters.

Some of this stuff does not depend on the image data and can be cached, some needs to be done with the image.
* `nearest` has the fastest non-cacheable part (~50 ms on my system for 1 quadrant of a megapixel image), seems to be the least biased, but the noisiest.
* `linear` is about twice slower, less noisier, but produces somehow biased β in very-low-intensity parts.
* `remap` has the best total time, but a slightly worse throughput (non-cacheable) than `nearest`, has the lowest noise on the peaks, but behaves crazily in very-low-intensity parts.

So it all seems to work _in principle_, but needs some improvements before adding to PyAbel:
1. Now it works with 1 quadrant only. This should be extended to full and half (?) images. Maybe by automatic folding (fast, but limited), maybe by actually going through all the supplied data. What do other `polar`/`vmi` tools expect in this regard?
2. It probably should be integrated deeply with the available tools. On the one hand, `remap` relies on the same thing as `reproject_image_into_polar`, but cannot reuse it directly. On the other hand, `angular_integration` and `anisotropy_parameter` can be reimplemented as particular simplified cases of these anisotropy-distribution calculations.
3. The sin θ weighting should be added (optional, used by default?).
4. Specifying the radial range (possibly beyond clipping) would be useful.
5. Specifying the radial step could be useful? Or not really? (This has potential problems of having less pixels than parameters.)
6. It can be relatively easily extend to arbitrary orders. Maybe, including odd orders (unlike the even-orders decomposition, they are not strict, but still useful for alignment studies).
7. (Optionally) clip the β values to their meaningful ranges?
8. Consider the caching API. It can be made like in the transform methods, but here we do not have a separate module.

Any comments, suggestions, concerns, considerations?

**Comment by MikhailRyazanov** [comment]
Ah, I think, there must be currently missing _sothething_-π factor (4π or 4/3 π or something else) for the intensity. And maybe a different _r_ scaling.

**Comment by stggh** [comment]
@MikhailRyazanov this is very neat and very clever. Thanks!
1. `abel.tools.vmi` operates on an individual quadrant (vertically flipped from your implementation), or the whole image, depending on the image `origin` parameter.
3. The intensity variation for photoelectron images is <math>&sigma;/4&pi; [ 1 + &beta;P<sub>2</sub>(cos&theta;)]</math>. There is no <math>sin&theta;</math> term.
4. and 7.  The common technique is to plot/present  `I * beta` to hide the low signal issues = bad <math>&beta;</math> values.

Still playing ...

**Comment by MikhailRyazanov** [comment]
@stggh,
2. |sin θ| should be the weight in the fitting procedure (weighted least squares), since there is proportionally less signal near the axis. The fitting function is, of course, the same as you wrote.
7. `I * beta` is the 2nd spherical harmonic, which by itself does not tell much about the anisotropy. It can be meaningful if plotted together with `I` (the 0th harmonic) for comparison, but then plotting the cos² and sin² parts probably makes more sense (they both are at least positive and directly correspond to the “parallel” and “perpendicular” components).

**Comment by MikhailRyazanov** [comment]
The main thing about computations that I actually wanted to ask:

Let's say, I have the 2D image array `IM` with pixel values and an array `bin` of the same dimensions, which contains integers indicating to which radial bin (1D) each pixel corresponds. So I need to bin the images pixels like this:
```
res = np.array(...)
for i in range(height):
  for j in range(width):
    b = bin[i, j]
    res[b] += IM[i, j]
```
(notice the `+=` operator — there are many pixels that must be summed into a bin).

This code is very slow, since it runs in the Python interpreter. Is there a fast way to perform this kind of operation in `numpy`/`scipy`/whatever?

(Currently I'm doing this backwards by precomputing lists of image pixel indices corresponding to each bin and then iterating over the bins and using `IM[indices[b]].sum()`. It is faster than the loops above, but looks like a wrong way.)

**Comment by stggh** [comment]
Isn't the/your `remap` method the answer?

**Comment by MikhailRyazanov** [comment]
The idea is that `nearest` and `linear` use each image pixel exactly once. `remap` distorts and interpolates the data. And `scipy.ndimage.map_coordinates` cannot add pixel values.

However, I've found the solutions myself:
* [`scipy.ndimage.sum`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.sum.html) is designed do this kind of binning.
* But also [`numpy.bincount`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html) can be used as `numpy.bincount(bin.reshape(-1), IM.reshape(-1))`, with the image pixels as weights... Might appear unusual, but this is actually the fastest method.

The [updated version](https://gist.github.com/MikhailRyazanov/8092647cbf3a4d5efc24e5120d364fd6/aff5fc452e37ea09bf1e6cdcabdfda1cc274ea5e) works much faster and is much simpler, as expected. `nearest` now can process megapixel images in just 4 ms!

And the intensity indeed should be 4π _r_<sup>2</sup> _P_<sub>0</sub>.

The Dribinski example again:
![drib_new](https://user-images.githubusercontent.com/12854133/58210705-6658f780-7ca7-11e9-9876-1509092731fb.png)
Looks pretty much as in the article.

---
@DanHickstein, there is a mistake in `abel.tools.analytical.SampleImage._dribinski()`:
```
t9 = 20*_gauss(r, 45, 3600)  # background under t3 to t5
```
must be
```
t9 = 20*_gauss(r, 45, 60)  # background under t3 to t5
```
(the last parameter is squared inside `_gauss()`).

Can you please correct it directly in the master branch? I don't think that we need a PR for this...

**Comment by stggh** [comment]
This looks good. Here is the `O2-ANU1024.txt.bz2' test case:
![image](https://user-images.githubusercontent.com/10932229/58217554-a622ec00-7d46-11e9-9829-a5c0a0556126.png)

[code gist](https://gist.github.com/stggh/b9e5010552efa9ce16ee93798c33a4f2)  (lacks comments)


You are right, there is a case for filtering the <math>&beta;</math> values.

**Comment by MikhailRyazanov** [comment]
When β varies across the peak, it might be an indication of some image distortions or imperfect centering (I had [some experience](https://www.tandfonline.com/doi/abs/10.1080/00268970802245192) ;–) with this), unless there is a good physical reason. And if it is not expected to vary, then getting it from intensity components radially integrated over the peak is the most proper way.
However, the strange thing in your plot is that `linbasex` and image fitting yield opposite trends.

**Comment by stggh** [comment]
> strange thing in your plot is that linbasex and image fitting yield opposite trends.

This represents the noise, and perhaps, a displacement between the transform methods.  Extending the radial-range around each peak (<math>&pm; 2&Gamma;</math>) is more illustrative.  Some <math>&beta;</math> smoothing required.

![image](https://user-images.githubusercontent.com/10932229/58226148-b9df4a00-7d68-11e9-8609-1e86dac912c6.png)

`linbasex` has a `smoothing` parameter (> 1 pixel). For `smoothing=2`:

![image](https://user-images.githubusercontent.com/10932229/58226327-9537a200-7d69-11e9-9a00-180779652c54.png)

**Comment by stggh** [comment]
I should add that any smoothing (> 1 pixel) degrades the structure in the (my) radial intensity spectrum, so for `linbasex` the two cannot be used/compared together.

**Comment by DanHickstein** [comment]
Wow! This looks super awesome! Thanks Mikhail!

> 1. Now it works with 1 quadrant only. This should be extended to full and half (?) images. Maybe by automatic folding (fast, but limited), maybe by actually going through all the supplied data. What do other polar/vmi tools expect in this regard?

As @stggh mentions, the other tools can set the origin arbitrarily, so can work with either the whole image or a quadrant. But, the sort of assume the full image. I don't think that we have good precedent either way, and would welcome a good suggestion on how to handle this. 

> 2. It probably should be integrated deeply with the available tools. On the one hand, remap relies on the same thing as reproject_image_into_polar, but cannot reuse it directly. On the other hand, angular_integration and anisotropy_parameter can be reimplemented as particular simplified cases of these anisotropy-distribution calculations.

Yes, we should integrate it as well as possible. But, the angular integration function is currently reasonably fast, so if this method is significant slower (it's probably not, but we should check), we may want to retain both methods. 

> 4. Specifying the radial range (possibly beyond clipping) would be useful.

Agreed.

> 5. Specifying the radial step could be useful? Or not really? (This has potential problems of having less pixels than parameters.)

I would probably allow the radial step to be specified, with a good description in the docstring that it's probably not a good idea to set it too much below it's default value of 1 pixel.

> 6. It can be relatively easily extend to arbitrary orders. Maybe, including odd orders (unlike the even-orders decomposition, they are not strict, but still useful for alignment studies).

Interesting! That sounds useful for some people.

> 7. (Optionally) clip the β values to their meaningful ranges?

Yeah, we can either add a functionality to do this, or (easier) include an example in the docstring showing people how to set the beta values to np.nan for ranges where the intensity is low.

> 8. Consider the caching API. It can be made like in the transform methods, but here we do not have a separate module.

I don't understand this comment - what are you trying to cache?

> Can you please correct it directly in the master branch? I don't think that we need a PR for this...

I think that all changes to the code should be accompanied by a PR. Then, we can always look back and see when a change occurred and (if we are doing our job right) why it occurred. Also, you found the bug, so you should get credit for fixing it.

**Comment by MikhailRyazanov** [comment]
@DanHickstein, thanks!

<p>1. Now I have the folding working, need to test whether it is faster than just going over the whole image. But yes, specifying an arbitrary origin works and is the universal way.

<p>8. The calculations involve some parts that do not depend on the actual image (if you look at the printout from the example, these are the lines that do not have “IMG”), so they can be stored and reused for repetitive analysis of multiple images (just like the transform methods now cache the transform matrices).  In some cases this increases the throughput significantly.

I am currently thinking about making this an object, which will store the parameters (and all precalculations), then you can feed it with images and get any desired distributions. But maybe several separate functions will be enough.

<p>6. Even orders are in progress.

When I'm done with this, I plan to make a PR. And leave 5., 7., and odd orders from 6. for the next iteration(s).

>> Can you please correct it directly in the master branch? I don't think that we need a PR for this...
>
> I think that all changes to the code should be accompanied by a PR. Then, we can always look back and see when a change occurred and (if we are doing our job right) why it occurred. Also, you found the bug, so you should get credit for fixing it.

OK, then I'll prepare it soon...

**Comment by DanHickstein** [comment]
> OK, then I'll prepare it soon...

You can also make a quick PR just for the bug fix and we can merge it right away...

**Comment by MikhailRyazanov** [comment]
Well, since most of the ideas here are implemented in #257 and #266, I'm closing this issue.

The remaining possible improvements are the adjustable radial step and clipping the β values. If somebody really needs them, it would be better to open separate issues.
